### ISSUE TITLE
RFE: option for "ftraceroute" functionality

### ISSUE THREAD

**Issue by jccleaver** [schweikert/fping#71]
After a bit of a search, it seemed like there was nothing out there that could multiplex traceroutes and let us do them in parallel, but on further reflection it feels like this is something fping could be bent into doing somewhat. We've already got the ability to specify TTL, and the parallel RR of responses is do-able. New record structures would need to be made, and output formats decided on (prefix each line with ultimate destination and output as received? Or wait until completed and output blocks? etc...), but nothing too horribly insurmountable.

This would come in very handy for our monitoring system, which uses fping for massively parallel pinging, but ends up doing serial traceroutes on things that are unreachable, which obviously slows things down.

I'd be willing to help work on a patch, but I wanted to put this RFE out there first :)

**Comment by hmh** [comment]
An ftraceroute would be very welcome, even if it starts as ICMP-based-only...

**Comment by gsnw-sebast** [comment]
I will try to tackle this function. As a first step, I will create and test a parallel test function outside of main_loop().

**Comment by auerswal** [comment]
I have seen mention of existing tools that allow traceroute to multiple targets, but I have not (yet) looked closer at them, and did not (yet) try them out:

* https://trippy.rs/
* https://github.com/lance0/ttl

(Both are written in Rust.)

**Comment by hmh** [comment]
Usually rust results in large binaries (just like golang).  Mostly irrelevant for desktops and laptops, but prohibitive in embedded (I mean constrained devices like Wifi routers, APs, switches, etc).  I.e. exactly those where you cannot just run 30 traceroutes in parallel due to RAM usage and would like a ftraceroute instead... ;-)

**Comment by gsnw-sebast** [comment]
I just published my traceroute test code under https://github.com/gsnw/ftraceroute . My pull request for fping is partly based on this experience.

I already have a version with pthread, which allows traceroute to run in parallel with multiple hostnames.

**Comment by schweikert** [comment]
I don't think we should use pthread. Really, I am not sure if we need this functionality, especially if it increases complexity significantly. Using pthread would be a big change for fping, which is single-threaded and using a select loop.

**Comment by gsnw-sebast** [comment]
I didn't use pthread in the pull request. It also works with select loop.

As an example, I created a standalone ftraceroute in my repository. I also think that fping shouldn't include a traceroute function.

Especially since quite a bit of code would have to be added to make the output look nicer.

**Comment by hmh** [comment]
I agree that fping should not try to traceroute, and I should have said as much when I commented that "ftraceroute would be nice".

tracerouting should be done in a separate tool, and it is a complex topic in itself (refer to paris-traceroute, dublin-traceroute, etc).

I have no idea if it would make sense to even share the same source tree with fping.
